pages/api/stats/index.ts

```diff
@@ -17,16 +20,55 @@ import withAuth, { NextApiRequestWithAuth } from '../../../lib/withAuth';
 import Level, { EnrichedLevel } from '../../../models/db/level';
 import Record from '../../../models/db/record';
 import Stat from '../../../models/db/stat';
-import { LevelModel, PlayAttemptModel, RecordModel, StatModel, UserModel } from '../../../models/mongoose';
-import { queueRefreshIndexCalcs } from '../internal-jobs/worker';
+import { AchievementModel, LevelModel, PlayAttemptModel, RecordModel, StatModel, UserModel } from '../../../models/mongoose';
+import { queueRefreshAchievements, queueRefreshIndexCalcs } from '../internal-jobs/worker';
 import { matchMarkCompleteLevel } from '../match/[matchId]';
 
-export async function issueAchievements(userId: Types.ObjectId, score: number, options: SaveOptions) {
-  for (const achievementType in AchievementInfo) {
-    const achievementInfo = AchievementInfo[achievementType];
+/**
+*  Creates a new achievement for the user
+ * @param userId
+ * @throws Error if user not found
+ */
+export async function refreshAchievements(userId: Types.ObjectId, categories: AchievementCategory[]) {
+  // it is more efficient to just grab all their achievements then to loop through and query each one if they have it
+  const [user, levelsCompletedByDifficulty, userCreatedLevels, allAchievements] = await Promise.all([
+
+    /** TODO: conditionally fetch these based on categories since not all categories need all this data
+     * One way to do this is to grab the function defintion of the `unlocked` function and see what data it needs
+     * Another way to is to define in the categories what data is needed and merge it in an object
+     */
+    UserModel.findById<User>(userId, {}, { lean: true }),
+    getCompletionByDifficultyTable(userId),
+    LevelModel.find<Level>(
+      {
+        userId: userId, isDraft: false, isDeleted: { $ne: true },
+      }, { score: 1, authorNote: 1, leastMoves: 1, ts: 1, calc_reviews_score_laplace: 1, calc_playattempts_just_beaten_count: 1, calc_playattempts_unique_users: 1 },
+      { lean: true }),
+    AchievementModel.find<Achievement>({ userId: userId }, { type: 1, }, { lean: true }),
+
+  ]);
+
+  if (!user) {
+    throw new Error('User not found');
```

Comment:
Is throwing an exception necessary? We can maybe just return null so we can avoid try/catch every time this function is used

Reply:
Good point